### PR TITLE
feat(ledger): governance ratification, tally, and enactment

### LINKED ISSUE #1955: governance: complete CIP-1694 phases 5-8 (tallying, ratification, enactment, epoch boundary)
## Summary

CIP-1694 governance is partially implemented in Dingo. Phases 1-2 are complete (data structures, DRep management, voting power calculation, activity/expiry tracking, voter type mapping — PRs #1329, #1373). Phases 3-4 are partial (CC hot key auth and resignation done; CC member expiry, threshold calculations, no-confidence handling, parent action chain validation, action expiry not yet). Phases 5-8 are not started.

This issue tracks the remaining work needed for Conway-era governance compliance: vote tallying, ratification, enactment, and epoch boundary processing. The gouroboros conformance suite for these areas (314 Conway tests including GOV/GOVCERT/ENACT/RATIFY/EPOCH) already passes for the protocol layer, so the validation rules are well-defined.

## Phase 5 — Vote Tallying

- DRep vote tallying against active stake distribution
- SPO vote tallying against active stake distribution
- CC vote tallying with member quorum
- Threshold checking against per-action thresholds defined in protocol parameters

Voting thresholds (already defined in protocol params):
- DRep: `dvtCommitteeNoConfidence`, `dvtCommitteeNormal`, `dvtHardForkInitiation`, `dvtPPEconomicGroup`, `dvtPPGovGroup`, `dvtPPNetworkGroup`, `dvtPPTechnicalGroup`, `dvtTreasuryWithdrawal`, `dvtUpdateToConstitution`
- SPO: `pvtCommitteeNoConfidence`, `pvtCommitteeNormal`, `pvtHardForkInitiation`, `pvtPPSecurityGroup`

## Phase 6 — Ratification & Enactment

- Detect when an action has crossed all required thresholds (DRep + SPO + CC where applicable)
- Enact `ParameterChange`
- Enact `UpdateCommittee` (add/remove CC members)
- Enact `NewConstitution` (update constitution hash)
- Enact `TreasuryWithdrawal` (move funds from treasury)
- Schedule `HardForkInitiation` for the next epoch boundary

## Phase 7 — Certificate Processing (Completion)

- Wire all governance certificate types into the existing ledger certificate pipeline (most done; verify completeness for `RegPool`, `UnregDRep`, vote/stake delegation combos)
- Voting procedure processing (vote messages on actions)
- Proposal procedure processing (action submission with deposit)
- Parent action chain validation: previous action root tracking. Requires a `gov_roots` table not yet created.

## Phase 8 — Epoch Boundary Processing

- Wire ratification + enactment into the epoch transition code path (`event/epoch.go` `EpochTransitionEvent`)
- Recalculate tallies at boundary
- Process ratifications discovered during the epoch
- Enact actions
- Expire old / unratified actions and return their deposits
- CC member expiry tracking (currently CC member expiry epoch is not stored in the DB — needs schema + migration)

## Action Types Reference

| Action | CC Vote | DRep Vote | SPO Vote |
|--------|---------|-----------|----------|
| NoConfidence | - | yes | yes |
| UpdateCommittee | - | yes | yes |
| NewConstitution | yes | yes | - |
| HardForkInitiation | yes | yes | yes |
| ParameterChange | yes | yes | varies |
| TreasuryWithdrawal | yes | yes | - |
| Info | - | yes | yes |

## Testing

- Unit tests for tallying, threshold checking, ratification detection per action type
- Integration tests for full proposal → vote → ratify → enact flow per action type
- Test edge cases: tied votes, threshold-exactly-met, expired actions, withdrawn actions
- Test against real mainnet data once enactment lands

## Roadmap

Q2 2026 — testnet block production prerequisites; Conway-era compliance.

### PR DESCRIPTION
**Pull request by wolf31o2**
Closes #1955 

---
## Summary by cubic
Implements the full CIP‑1694 governance tick at epoch rollover to ratify, tally, enact, and expire proposals with rollback‑safe refunds and persisted committee quorum. Adds a dedicated `ledger/governance` pipeline integrated with chainsync, deltas, backfill, `view`, rollback, and snapshot import.

- **New Features**
  - Epoch tick: enact previous‑epoch ratified, expire stale (`expired_epoch`/`expired_slot`), then tally/ratify; enforce parent‑chain roots, single ratification per purpose, deterministic ordering.
  - Enactment and refunds: apply protocol params and committee + quorum; credit deposit refunds to reward accounts via a journal; rollback removes post‑slot reward credits and committee updates.
  - Voting/tally: DRep voting power now = live UTxO + reward balances; batch by credential and predefined types; detect resigned members; bootstrap/post‑bootstrap CC/SPO/DRep rules.
  - Storage/API: persist `gov_action_cbor`; ordered active/expiring proposal queries; store committee quorum; normalize committee hot key to `HotCredential`; `view` returns seated members with hot keys; moved `ProcessProposals`/`ProcessVotes` into `ledger/governance`; snapshot import ingests committee and quorum.

- **Migration**
  - Run DB migrations to add `gov_action_cbor`, `expired_epoch`/`expired_slot`, `drep_type`, `committee_quorum`, and the `account_reward_delta` journal; enable committee soft‑delete. Restart the node.

<sup>Written for commit 28e3a90433de64fb90f33df417a82ffe95856659. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Full governance pipeline: enactment, expiry, ratification with deterministic ordering; proposal payload persistence.
  * Committee & quorum management: persisted quorum, resigned-member detection, soft-delete/include-deleted views, rollback-aware restoration.
  * Voting & rewards: DRep voting-power now includes reward balances with batch/by-type queries; account reward journaling with rollback.
  * DB migrations: new reward-journal and committee-quorum tables.

* **Refactor**
  * Governance logic moved into a dedicated package and integrated into epoch/chain processing.

* **Tests**
  * Expanded unit/integration coverage across enactment, ratify/tally, quorum, soft-delete, rewards, and rollback.

### CODE REVIEW

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

File: ledger/governance/epoch.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+		n = 8
+	}
+	return hex.EncodeToString(h[:n])
+}
+
+// defaultCCQuorum is the bootstrap fallback quorum used until the
+// committee's configured quorum is modeled and persisted. Matches the
+// Conway genesis default so we don't silently auto-approve every
+// CC-gated action.
+var defaultCCQuorum = big.NewRat(2, 3)
```
Comment (agaffney) [inline ledger/governance/epoch.go:507]: Why not read this from the Conway genesis config?

File: ledger/governance/tally.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	members, err := ctx.DB.GetActiveCommitteeMembers(ctx.Txn)
+	if err != nil {
+		return fmt.Errorf("get active cc members: %w", err)
+	}
+	tally.CCTotalCount = len(members)
+
+	activeHotKeys := make(map[string]struct{}, len(members))
+	for _, m := range members {
+		activeHotKeys[string(m.HostCredential)] = struct{}{}
+	}
```
Comment (agaffney) [inline ledger/governance/tally.go:360]: `HostCredential` should be renamed to `HotCredential` as noted by the bot

**Review by agaffney — DISMISSED** [review]

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

File: database/models/account.go [hunk 3]
```diff
@@ -24,6 +24,28 @@ import (
 
 var ErrAccountNotFound = errors.New("account not found")
 
+const (
+	DrepTypeAddrKeyHash        uint64 = 0
```
Comment (chrisguiney) [inline database/models/account.go]: this should probably use `iota`

File: ledger/governance/epoch.go (hunk 1 above)
Comment (wolf31o2) [reply ledger/governance/epoch.go:507]: Done

File: database/models/account.go (hunk 3 above)
Comment (wolf31o2) [reply database/models/account.go]: Switched to iota in `database/models/account.go` — values stay 0..3 as the protocol requires.

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@coderabbitai review
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Comment by wolf31o2** [comment]
@cubic-dev-ai review

**Review by verbotenj — APPROVED** [review]
